lib/src/configuration.dart

```diff
@@ -394,14 +399,88 @@ class RealmSchema extends Iterable<SchemaObject> {
 
 /// The signature of a callback that will be invoked if a client reset error occurs for this [Realm].
 ///
-/// Currently, Flexible Sync supports only the [ManualSyncClientResetHandler].
-class SyncClientResetErrorHandler {
+/// A Client Resync is triggered if the device and server cannot agree
+/// on a common shared history for the Realm file,
+/// thus making it impossible for the device to upload or receive any changes.
+/// This can happen if the server is rolled back or restored from backup.
+/// {@category Sync}
+abstract class ClientResetHandler {
+
+  /// Defines what should happen in case of a Client Resync
+  ClientResyncModeInternal get _mode;
+
   /// The callback that handles the [SyncClientResetError].
-  final void Function(SyncClientResetError code) callback;
+  final ClientResetCallback callback;
 
-  /// Initializes a new instance of of [SyncClientResetErrorHandler].
-  const SyncClientResetErrorHandler(this.callback);
+  /// Initializes a new instance of of [ClientResetHandler].
+  const ClientResetHandler(this.callback);
 }
 
 /// A client reset strategy where the user needs to fully take care of a client reset.
-typedef ManualSyncClientResetHandler = SyncClientResetErrorHandler;
+///
+/// If you set [ManualRecoveryHandler] callback as `clientResetHandler` argument of [Configuration.flexibleSync],
+/// that will enable full control of moving unsynced changes to synced realm.
+/// {@category Sync}
+class ManualRecoveryHandler extends ClientResetHandler {
+  const ManualRecoveryHandler({required ClientResetCallback onReset}) : super(onReset);
+
+  @override
+  ClientResyncModeInternal get _mode => ClientResyncModeInternal.manual;
+}
+
+/// A client reset strategy where all the not yet synchronized data is automatically
+/// discarded and a fresh copy of the synchronized Realm is obtained.
+///
+/// If you set [DiscardUnsyncedChangesHandler] callback as `clientResetHandler` argument of [Configuration.flexibleSync],
+/// the local Realm will be discarded and replaced with the server side Realm.
+/// All local changes will be lost.
+/// {@category Sync}
+class DiscardUnsyncedChangesHandler extends ClientResetHandler {
+  const DiscardUnsyncedChangesHandler({required ClientResetCallback onReset}) : super(onReset);
```

Comment:
I converted this PR to draft as there are more changes incoming. It was my suggestion to use named args because we'll have more callbacks there. The final signature will look something like `DiscardUnsyncedChangesHandler({BeforeCallback? beforeCallback, AfterCallback? afterCallback, ClientResetCallback? manualFallback})`. Having a function where you pass 3 optional callbacks without named args is going to look very confusing.

Reply:
Shall I merge https://github.com/realm/realm-dart/pull/970 to this PR?

Reply:
No, #970 is not ready at all yet. I'd like to get at least 1 test running to verify that things work as they should.

Reply:
OK. This PR was only for setting the default value of resync mode. Maybe it is better to combine both tasks in one PR.